### PR TITLE
Updated CMake structure for working install, and to use CMake build Q…

### PR DESCRIPTION
**Pull request by bjoo**
…IO and QMP

-- simplified linking against QIO and QMP.  It is difficult to use the existing download mechanism and also then set the approrpiate QIO_DIR and QMP_DIR. I would advocate either going the CMP route @mathiaswagner was thinking of or to use CMakeFetchContent for QIO and QMP and then get to them using `add_subdirectory`

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,10 +1,12 @@
 # basic setup for cmake
-cmake_minimum_required(VERSION 3.15 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
```
Comment (bjoo) [inline CMakeLists.txt:2]: I don' think this is too restrictive these days. @maddyscientist is using 3.20rc features

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 
 if(POLICY CMP0074)
   cmake_policy(SET CMP0074 NEW)
 endif()
 
+cmake_policy(SET CMP0060 NEW)
```
Comment (bjoo) [inline CMakeLists.txt]: I have removed this in an updated

File: CMakeLists.txt [hunk 3]
```diff
@@ -454,12 +456,7 @@ message(STATUS "CUDA Compiler is" ${CMAKE_CUDA_COMPILER})
 message(STATUS "Compiler ID is " ${CMAKE_CUDA_COMPILER_ID})
 
 # CUDA Wrapper for finding libs etc
-if(CMAKE_VERSION VERSION_LESS 3.17)
-  find_package(CUDAWrapper)
-else()
-  # for cmake 3.17+ we rely on
-  find_package(CUDAToolkit)
-endif()
+find_package(CUDAToolkit)
```
Comment (bjoo) [inline CMakeLists.txt]: Since we require minimum 3.17 here now, I could remove this if() to simplify code

File: CMakeLists.txt [hunk 4]
```diff
@@ -573,128 +570,33 @@ if(QUDA_QDPJIT)
...
-        "-lqmp"
-        CACHE STRING "LIBS for QMP - should be derived from qmp-config --libs")
-    ExternalProject_Add_Step(
-      QMP reconf
-      COMMAND ${AUTORECONF_EXE} -fi
-      WORKING_DIRECTORY <SOURCE_DIR>
-      DEPENDERS configure
-      DEPENDEES download)
-
+  message(STATUS "Looking for QMP")
```
Comment (bjoo) [inline CMakeLists.txt]: This is the modern cmake way to look for a QMPConfig.cmake. One either needs to set QMP_DIR to <install_path_for_QMP>/lib/cmake/QMP 
or add <install_path_for_QMP> to CMAKE_PREFIX_PATH

File: CMakeLists.txt [hunk 5]
```diff
@@ -573,128 +570,33 @@ if(QUDA_QDPJIT)
...
-    set(QUDA_QMP_LDFLAGS
-        ${QUDA_QMP_LDFLAGS_INTERNAL}
-        CACHE STRING "LDFLAGS for QMP - should be derived from qmp-config --ldflags")
-    set(QUDA_QMP_LIBS
-        ${QUDA_QMP_LIBS_INTERNAL}
-        CACHE STRING "LIBS for QMP - should be derived from qmp-config --libs")
+    message(STATUS "QMP not found")
+  endif()
+
+  if(QUDA_DOWNLOAD_USQCD AND NOT QMP_FOUND)
```
Comment (bjoo) [inline CMakeLists.txt:619]: find_package() won't find QMP until it is built. So we can't use it to find a dependency that is downloaded using `External_Project_Add`. We could use `FetchContent` to a subdir and add that subdir tho. Or we can use a package manager. So for now I disabled this

File: CMakeLists.txt [hunk 6]
```diff
@@ -573,128 +570,33 @@ if(QUDA_QDPJIT)
...
-  find_library(QIO_FOUND qio HINTS ${QUDA_QIOHOME}/lib NO_DEFAULT_PATH)
-  mark_as_advanced(QIO_FOUND)
-  set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${QIO_FOUND})
-  if(QUDA_DOWNLOAD_USQCD AND NOT QIO_FOUND)
-    set(QIO_BRANCH
-        "extended")
-    if(QUDA_DOWNLOAD_QIO_LEGACY)
-      set(QIO_BRANCH
-          "qio2-5-0")
+  message(STATUS "Looking for QIO")
```
Comment (bjoo) [inline CMakeLists.txt]: Similar behaviour to QMP

File: CMakeLists.txt [hunk 7]
```diff
@@ -877,3 +779,19 @@ include(CTest)
 add_subdirectory(lib)
 add_subdirectory(tests)
 add_subdirectory(doc)
+
+
+include(CMakePackageConfigHelpers)
+configure_package_config_file(QUDAConfig.cmake.in QUDAConfig.cmake
```
Comment (bjoo) [inline CMakeLists.txt:857]: CMake style is that package names start with a capital. In this case QUDA may be all in Capitals, or Quda depending on how one feels as long as one is consistent. I have gone with QUDA.

File: CMakeLists.txt [hunk 8]
```diff
@@ -877,3 +779,19 @@ include(CTest)
...
+	INSTALL_DESTINATION lib/cmake/QUDA)
+
+write_basic_package_version_file(
+	QUDAVersion.cmake
+	VERSION ${PACKAGE_VERSION}
+	COMPATIBILITY AnyNewerVersion)
+
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/QUDAVersion.cmake
+	${CMAKE_CURRENT_BINARY_DIR}/QUDAConfig.cmake
+	DESTINATION lib/cmake/QUDA)
```
Comment (bjoo) [inline CMakeLists.txt:867]: Best to have this the same capitalization as the package name

File: QUDAConfig.cmake.in [hunk 9]
```diff
@@ -0,0 +1,41 @@
+@PACKAGE_INIT@
```
Comment (bjoo) [inline QUDAConfig.cmake.in:1]: This is a very basic config file. It sets some flags from the existing configuraiton, and so depending on those we can resolve dependencies.

File: lib/CMakeLists.txt [hunk 10]
```diff
@@ -413,13 +413,11 @@ endif()
 
 if(QUDA_QMP)
   if(QUDA_DOWNLOAD_USQCD AND NOT QMP_FOUND)
-    add_dependencies(quda QMP)
-    add_dependencies(quda_cpp QMP)
+    add_dependencies(quda QMPBuild)
```
Comment (bjoo) [inline lib/CMakeLists.txt]: QMP is now the namespace for the imported QMP targets (QMP::qmp). So I could not leave these as 'QMP' but needed to rename them. We can use a different name if desired.

File: lib/CMakeLists.txt [hunk 11]
```diff
@@ -413,13 +413,11 @@ endif()
...
+    add_dependencies(quda_cpp QMPBuild)
   endif()
-  target_include_directories(quda SYSTEM PUBLIC $<BUILD_INTERFACE:${QUDA_QMPHOME}/include>)
   target_compile_definitions(quda PUBLIC QMP_COMMS)
-  target_link_libraries(quda INTERFACE ${QUDA_QMP_LDFLAGS} ${QUDA_QMP_LIBS})
-  target_link_libraries(quda PUBLIC MPI::MPI_CXX)
+  target_link_libraries(quda PUBLIC QMP::qmp)
```
Comment (bjoo) [inline lib/CMakeLists.txt:442]: The nice way to link QMP :D

File: lib/CMakeLists.txt [hunk 12]
```diff
@@ -428,9 +426,7 @@ if(QUDA_QIO)
     add_dependencies(quda_cpp QIO)
   endif()
   target_compile_definitions(quda PUBLIC HAVE_QIO)
-  target_include_directories(quda SYSTEM PUBLIC $<BUILD_INTERFACE:${QUDA_QIOHOME}/include>)
-  target_include_directories(quda SYSTEM PUBLIC $<BUILD_INTERFACE:${QUDA_LIMEHOME}/include>)
-  target_link_libraries(quda INTERFACE ${QUDA_QIO_LDFLAGS} ${QUDA_QIO_LIBS})
+  target_link_libraries(quda PUBLIC QIO::qio)
```
Comment (bjoo) [inline lib/CMakeLists.txt:460]: The nice way to link QIO

File: lib/CMakeLists.txt [hunk 13]
```diff
@@ -585,13 +581,20 @@ add_custom_command(
   PRE_LINK
   COMMAND ${CMAKE_COMMAND} -E remove ${CMAKE_CURRENT_BINARY_DIR}/libquda.a ${CMAKE_CURRENT_BINARY_DIR}/libquda.so)
 
+set_target_properties(quda PROPERTIES INSTALL_RPATH_USE_LINK_PATH ON )
```
Comment (bjoo) [inline lib/CMakeLists.txt]: I have removed this in a later version

File: lib/CMakeLists.txt [hunk 14]
```diff
@@ -585,13 +581,20 @@ add_custom_command(
...
+
 install(
   TARGETS quda
-  EXPORT qudaTargets
+  EXPORT QUDATargets
```
Comment (bjoo) [inline lib/CMakeLists.txt:615]: Went back to use same captialization as package name

Comment (bjoo) [reply lib/CMakeLists.txt:615]: I just rebased this with current `develop` to make it up-to-date with that. Includes @weinbe2 's most recent patch

**Comment by mathiaswagner** [comment]
I merged release/1.1 but we will only merge this for the QUDA 2.0 so need to combine it with the generic kernels branch at some point.

**Comment by mathiaswagner** [comment]
We now have #1156

**Comment by maddyscientist** [comment]
Closing this PR, since it's superseded by #1156

**Pull request closed without merging**